### PR TITLE
Define a feature configuration which apply to a controller action

### PR DESCRIPTION
**Pull request by jdecool**
This PR allow developer to define controller access by checking if a feature flag is enabled or not.

### With annotation

```php
class MyController extends Controller
{
    /**
     * @Feature("foo")
     * @Feature("foo", enabled=true)
     */
    public function annotationFooEnabledAction()
    {
        return new Response('MyController::annotationFooEnabledAction');
    }

    /**
     * @Feature("foo", enabled=false)
     */
    public function annotationFooDisabledAction()
    {
        return new Response('MyController::annotationFooDisabledAction');
    }
}
```

### In the routing configuration

```yaml
# app/config/routing.yml
my_first_route:
    path: /my/first/route
    defaults:
        _controller: AppBundle:Default:index
        _features:
            - { feature: my_feature_key, enabled: false } # The action is accessible if "my_feature_key" is disabled
            
my_second_route:
    path: /my/second-route
    defaults:
        _controller: AppBundle:Default:second
        _features:
            - { feature: foo } # The action is accessible if "foo" is enabled ...
            - { feature: bar, enabled: true } # ... and "bar" feature is also enabled
```

### CODE REVIEW

File: EventListener/ControllerListener.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+use Doctrine\Common\Util\ClassUtils;
+use Novaway\Bundle\FeatureFlagBundle\Annotation\Feature;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\FilterControllerEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+class ControllerListener implements EventSubscriberInterface
+{
+    /** @var Reader */
+    private $reader;
```
Comment (skwi) [inline EventListener/ControllerListener.php]: `$annotationReader` might be more explicit

File: EventListener/ControllerListener.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            if ($annotation instanceof Feature) {
+                $key = $annotation->getFeature();
+                if (isset($features[$key])) {
+                    throw new \UnexpectedValueException(sprintf('Feature "%s" is defined more than once for %s::%s', $key, $className, $controller[1]));
+                }
+
+                $features[$key] = [
+                    'feature' => $annotation->getFeature(),
+                    'enabled' => $annotation->getEnabled(),
+                ];
```
Comment (skwi) [inline EventListener/ControllerListener.php]: On [line 36](https://github.com/novaway/NovawayFeatureFlagBundle/pull/4/files#diff-50211849393457d7baea165ea5fddce1R36) the PHP 5.3 syntax is used for arrays. Here we have the 5.4 syntax. 

Either 5.3 compatibility is required and this instruction fails, or there is a code standard issue.

I suggest we update the `composer.json` file with the min PHP supported version to make this clear. My suggestion is to drop 5.3 compatibility.

File: README.md [hunk 3]
```diff
@@ -98,6 +98,54 @@ You can also check a flag in your templates:
...
+#### As a controller annotation
+
+You can also restrict a controller access with annotations :
+
+```php
+class MyController extends Controller
+{
+    /**
+     * @Feature("foo")
+     * @Feature("foo", enabled=true)
```
Comment (skwi) [inline README.md]: `UnexpectedValueException : Feature "foo" is defined more than once for MyController:: annotationFooEnabledAction` 😉

File: EventListener/ControllerListener.php (hunk 2 above)
Comment (jdecool) [inline EventListener/ControllerListener.php]: Done

File: Annotation/Feature.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        if (!isset($values['value'])) {
+            throw new \RuntimeException('Feature annotation value is required.');
+        }
+
+        $this->feature = (string) $values['value'];
+
+        $this->enabled = true;
+        if (isset($values['enabled'])) {
+            $this->enabled = (bool) $values['enabled'];
+        }
```
Comment (jdecool) [inline Annotation/Feature.php]:
```php
$this->enabled = !isset($values['enabled']) && (bool) $values['enabled'];
```

Comment (jdecool) [inline Annotation/Feature.php]: The correct line is :

```php
$this->enabled = !isset($values['enabled']) || (bool) $values['enabled'];
```

**Pull request merged**